flow/project.py

```diff
@@ -267,6 +268,25 @@ def _make_bundles(operations, size=None):
             break
 
 
+class _HooksDecorator:
+    def __init__(self):
+        self._operation_hooks = defaultdict(lambda: defaultdict(list))
+
+    def __getitem__(self, func):
+        return self._operation_hooks[func]
+
+    def __getattr__(self, name):
+        class _InstallHook:
+            def __init__(install_self, hook_func):
+                install_self.hook_func = hook_func
+
+            def __call__(install_self, func):
+                self._operation_hooks[func][name].append(install_self.hook_func)
+                return func
+
+        return _InstallHook
```

Comment:
Do we want to allow any attribute? I feel we should restrict it to those that are proper hook triggers.
```
@FlowProject.hook.in_the_middel(lambda: operation_name, jobs)
```
should raise a `FlowProjectDefinition` error or `AttributeError`.

Reply:
https://github.com/glotzerlab/signac-flow/blob/fcab466a238ecc6bff524393dbae3573b5da1f38/flow/hooks/__init__.py#L16-L28

The restriction to proper hook triggers appears here and raises an `Exception`. Maybe we should change the error type though? `AttributeError` might more intuitive than an `Exception`. 

Should we include more comments throughout the code? I also missed the portion of the code that caught invalid hook attributes.